lib/concurrent/struct/final_struct.rb

```diff
@@ -0,0 +1,122 @@
+require 'concurrent/struct/abstract_struct'
+require 'concurrent/errors'
+require 'concurrent/synchronization'
+
+module Concurrent
+
+  # An thread-safe, write-one variation of Ruby's standard `Struct`.
+  # Each member can have its value set at most one, either at construction
+  # or any time thereafter. Attempting to assign a value to a member
+  # that has already been set will result in a `Concurrent::ImmutabilityError`.
+  #
+  # @see http://ruby-doc.org/core-2.2.0/Struct.html Ruby standard library `Struct`
+  # @see http://en.wikipedia.org/wiki/Final_(Java) Java `final` keyword
+  module FinalStruct
+    include AbstractStruct
+
+    # @!macro struct_values
+    def values
+      synchronize { ns_values }
+    end
+    alias_method :to_a, :values
+
+    # @!macro struct_values_at
+    def values_at(*indexes)
+      synchronize { ns_values_at(indexes) }
+    end
+
+    # @!macro struct_inspect
+    def inspect
+      synchronize { ns_inspect }
+    end
+    alias_method :to_s, :inspect
+
+    # @!macro struct_to_h
+    def to_h
+      synchronize { ns_to_h }
+    end
+
+    # @!macro struct_get
+    def [](member)
+      synchronize { ns_get(member) }
+    end
+
+    # @!macro struct_equality
+    def ==(other)
+      synchronize { ns_equality(other) }
+    end
+
+    # @!macro struct_each
+    def each
+      return enum_for(:each) unless block_given?
+      synchronize { ns_each(&Proc.new) }
```

Comment:
This is quite obscure, why not explicit `&block` argument? I think it would be more readable.